client/scripts/controllers/server/socket/chainEventsNs.ts

```diff
@@ -0,0 +1,68 @@
+import { WebsocketMessageType, WebsocketNamespaces } from 'types';
+import app from 'state';
+import { Notification, NotificationSubscription } from 'models';
+import { io } from 'socket.io-client';
+
+export class ChainEventsNamespace {
+  private ceNs;
+  private _isConnected = false;
+
+  constructor() {
+    this.ceNs = io(`/${WebsocketNamespaces.ChainEvents}`, {
+      transports: ['websocket'],
+    });
+    this.ceNs.on('connect', this.onconnect.bind(this));
+    this.ceNs.on('disconnect', this.ondisconnect.bind(this));
+    this.ceNs.on(
+      WebsocketMessageType.ChainEventNotification,
+      this.onChainEvent.bind(this)
+    );
+  }
+
+  public addChainEventSubscriptions(subs: NotificationSubscription[]) {
+    if (this._isConnected) {
+      const eventTypes = subs.map((x) => x.ChainEventType?.id).filter((x) => !!x);
+      console.log('Adding Websocket subscriptions for:', eventTypes);
+      this.ceNs.emit('newSubscriptions', eventTypes);
+    } else {
+      console.log('ChainEventsNamespace is not connected');
+    }
+  }
+
+  public deleteChainEventSubscriptions(subs: NotificationSubscription[]) {
+    if (this._isConnected) {
+      const eventTypes = subs.map((x) => x.ChainEventType?.id).filter((x) => !!x);
+      console.log('Deleting Websocket subscriptions for:', eventTypes);
+      this.ceNs.emit(
+        'deleteSubscriptions',
+        subs.map((x) => x.ChainEventType?.id)
+      );
+    } else {
+      console.log('ChainEventsNamespace is not connected');
+    }
+  }
+
+  private onChainEvent(notification: any) {
```

Comment:
Would be really cool to have proper types here -- could put them in `shared/` if we have them on the server.